### PR TITLE
fix: skip chartification when chart renders zero resources

### LINKED ISSUE #1757: Cannot apply transformers if Helm chart is empty
### Operating system

macOS Sonoma

### Helmfile Version

 1.0.0-rc.7

### Helm Version

v3.10.1

### Bug description

We have a number of charts that install addons into various clusters. Some of these charts render no resources for clusters for which we don't want to install the corresponding addon. However, for such charts, transformers fail to be applied with the following error:

```
assertion failed: unexpected dir entry "" it must be the abs path to the output directory
```

This appears to be a bug in chartify: https://github.com/helmfile/chartify/blob/6fbcd045d33e638ed6453b62bb2f044af5604e52/replace.go#L144

### Example helmfile.yaml

```yaml
releases:
- name: test
  chart: ./chart
  transformers:
  - apiVersion: builtin
    kind: LabelTransformer
    metadata:
      name: unused
    labels:
      foo: bar
    fieldSpecs:
    - path: spec/groups/rules/labels
      create: true
```

The `chart` directory contains the following files:

Chart.yaml:

```yaml
apiVersion: v2
type: application
name: test
version: 0.1.0
```

values.yaml

```yaml
enabled: false
```

templates/configmap.yaml

```yaml
{{- if .Values.enabled }}
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
{{- end }}
```

I then get the following error:

```sh
% helmfile template -f config/helmfile/helmfile.yaml
in config/helmfile/helmfile.yaml: [assertion failed: unexpected dir entry "" it must be the abs path to the output directory]
```

Removing the transformer works as expected

Changing `enabled: true` in values.yaml also works as expected

### Error message you've seen (if any)

Provided above

### Steps to reproduce

Provided above

### Working Helmfile Version

N/A

### Relevant discussion

_No response_

### PR DESCRIPTION
**Pull request by ankit090701**
## Summary

Helmfile crashes when a release has `transformers`, `jsonPatches`, or `strategicMergePatches` configured and its chart renders zero resources (e.g. everything gated behind a falsy `{{- if .Values.enabled }}`):

```
assertion failed: unexpected dir entry "" it must be the abs path to the output directory
```

## Root cause

`processChartification` in `pkg/state/state.go` calls into `github.com/helmfile/chartify`, which runs `helm template --output-dir <dir>` and then expects to find **exactly one** directory entry under `<dir>` (the rendered chart's output). See [`replace.go`](https://github.com/helmfile/chartify/blob/v0.28.0/replace.go#L130-L152) in chartify v0.28.0 (the version this repo currently depends on).

When the chart renders zero resources, `helm template --output-dir` produces an empty directory, so the loop over directory entries never runs, `chartOutputDir` is left as `""`, and chartify's own `filepath.IsAbs(chartOutputDir)` assertion fails on that empty string. chartify doesn't expose a typed/sentinel error for this specific case, so this is only observable by matching on the assertion text.

## Fix

Since there's nothing to chartify when a chart renders no resources, treat this specific chartify error as a no-op: fall back to using the (already dependency-rewritten) chart path as-is, exactly like the non-chartified local-chart path already does (`buildDeps = !skipDeps`, matching the existing convention at the `processLocalChart` call site). `helm template`/`upgrade`/etc. then run against the original chart directly, producing the same empty output helm would have produced without chartify involved at all. Any other chartify error continues to be surfaced unchanged — only this specific, distinctive assertion message is matched.

I deliberately didn't try to fix this in `chartify` itself, even though that's arguably the more "proper" root-cause location, since a fix there wouldn't take effect here until a new chartify version is released and this repo's `go.mod` is bumped to it — this change gets a working fix out now, entirely within this repo.

## Prior art

There was an earlier attempt at this exact fix in #2306, which took a similar approach (catching the chartify error and no-op'ing) but went stale waiting on a DCO sign-off from the contributor and was closed by stale-bot — it was never actually rejected on technical grounds. This PR is a fresh implementation (smaller diff, reuses the existing `!skipDeps` convention for `buildDeps`, adds a dedicated helper function with its own unit tests), signed off per `CONTRIBUTING.md`.

## Test plan

- [x] `go build ./...`
- [x] Added `pkg/state/issue_1757_test.go` with unit tests for the new `isChartifyEmptyRenderOutputError` helper: matches chartify's specific empty-output assertion, does **not** match chartify's distinct "more than one dir entry" assertion, an unrelated `kustomize not found` error, or a generic helm template error.
- [x] `go test ./pkg/state/...` passes (pre-existing unrelated failures on my Windows dev box in `TestLocalChartWithTransformersPathNormalization` / `TestChartPathNormalizationBeforeChartification` reproduce identically on unmodified `main` via `git stash`, confirmed unrelated to this change — looks like a `normalizeChart` Windows path-separator issue)
- [x] Manually verified end-to-end with real `helm` + `kustomize` binaries, using a local chart with `{{- if .Values.enabled }}` gating all templates and an `AnnotationsTransformer` configured:
  - `enabled: false` → previously crashed with the assertion error; now `helmfile template` exits 0, logs `chart rendered no resources, skipping chartification`, and renders the (empty) chart normally
  - `enabled: true` → still correctly applies the transformer (verified the annotation shows up in the rendered output), confirming the normal chartify path is unaffected by this change

Fixes #1757

### CODE REVIEW

**Comment by yxxhero** [comment]
Thanks for this — clean, well-scoped fix for #1757. I verified the core approach and have one blocking issue plus a few smaller notes.

## What I verified

- The matched substring against the actual dependency source: `chartify@v0.28.0/replace.go:151` emits exactly `assertion failed: unexpected dir entry %q it must be the abs path to the output directory`. The chosen substring `it must be the abs path to the output directory` is **distinctive** — it does not collide with the sibling assertions at `replace.go:140` (`...it must be a dir but was not`) or `replace.go:144` (`...there should be only one dir entry...`). The test even asserts the non-match with the latter. 👍
- `chartOutputDir` can only be `""` when the render is truly empty (the loop runs zero times), and `helmOutputDir` is absolute, so this assertion branch is *only* reachable via the empty-render case. The match is sound, not just lucky.
- `buildDeps = !skipDeps` is consistent with both the chartify success path and the `processLocalChart` branch. Good.
- Test file naming (`issue_1757_test.go`) and table-driven style match repo conventions.

## 🔴 Blocking — the no-op path can return a temp dir that's deleted by a deferred cleanup

`processChartification` rewrites relative `file://` deps into a temp copy before calling chartify:

```go
if st.fs.DirectoryExistsAt(chartPath) {
    rewrittenPath, cleanupTempChart, err := st.rewriteChartDependencies(chartPath)
    ...
    chartPath = rewrittenPath          // <-- chartPath is now a TEMP dir
    defer cleanupTempChart()           // <-- which gets RemoveAll'd on return
}
...
out, err := c.Chartify(...)
if err != nil {
    if isChartifyEmptyRenderOutputError(err) {
        ...
        return chartPath, !skipDeps, nil   // ⚠️ returns the temp dir
    }
    return "", false, err
}
chartPath = out   // success path is safe: chartPath is reassigned to chartify's own output
```

On the **success** path this is fine — `chartPath` is reassigned to `out` (chartify's output), so the deferred cleanup deleting the deps-rewrite temp is harmless. But the new **no-op** path returns `chartPath` *before* that reassignment, i.e. it returns the deps-rewrite temp dir, and the `defer` then immediately `RemoveAll`s it. The caller (state.go) hands that path to the subsequent `helm template`/`upgrade`, which then fails with "chart not found".

Trigger conditions: a **local chart directory** whose `Chart.yaml` has relative `file://` dependencies **and** renders zero resources **and** has a transformer/patch configured. Narrow, but real and reproducible — and notably *not* covered by the added unit test (which only exercises the pure string-matching helper).

### Suggested fix

Return the *original* (pre-rewrite) path, which is correct because the rewrite only exists to accommodate chartify's temp dir; without chartify, relative `file://` deps resolve fine from the original location. While here, also flatten the nested `if` into an early return (the `if err != nil { if isChartify... { } }` nesting isn't necessary and is slightly harder to read):

```go
func (st *HelmState) processChartification(...) (string, bool, error) {
    originalChartPath := chartPath
    if st.fs.DirectoryExistsAt(chartPath) {
        rewrittenPath, cleanupTempChart, err := st.rewriteChartDependencies(chartPath)
        if err != nil {
            return "", false, fmt.Errorf("failed to rewrite chart dependencies: %w", err)
        }
        chartPath = rewrittenPath
        defer cleanupTempChart()
    }
    ...
    out, err := c.Chartify(release.Name, chartPath, chartify.WithChartifyOpts(chartifyOpts))

    // Early-return the specific empty-render case — avoids nesting under `if err != nil`.
    if err != nil && isChartifyEmptyRenderOutputError(err) {
        st.logger.Debugf("release %q: chart rendered no resources, skipping chartification: %v", release.Name, err)
        // Return the original path, NOT the deps-rewritten temp copy (the deferred
        // cleanup would remove it). Relative file:// deps resolve from the original
        // location since chartify is not involved.
        return originalChartPath, !skipDeps, nil
    }
    if err != nil {
        return "", false, err
    }

    chartPath = out
    ...
}
```

This works for both local (already normalized to absolute by the caller) and remote charts (where the rewrite block is skipped and `originalChartPath == chartPath`).

## 🟡 Minor

1. **Test coverage gap.** The helper is well-tested, but nothing exercises the *wiring* inside `processChartification`, so the bug above (and any future regression in how the no-op path constructs its return) wouldn't be caught. The existing `issue_2297/2444/2355_test.go` files show the repo has patterns for testing this area; an integration-style test that injects/forces the empty-render condition (even by stubbing the chartify result) would materially de-risk this change.

2. **Upstream tracking.** The text-match is acknowledged and well-documented, and the rationale for not fixing in chartify (release velocity) is reasonable. That said, chartify returning a *typed/sentinel error* (or, better, a no-op success) for empty renders would be the proper fix. Suggest opening an upstream issue/PR in `helmfile/chartify` and linking it in the constant's doc comment so the substring match can eventually be retired.

3. **CI note.** You flagged pre-existing Windows-only failures in two transformers tests. CI runs Linux so they likely won't block merge, but worth confirming CI is green before approval.

## Verdict

The core idea is correct and the error matching is verified sound against chartify v0.28.0. The no-op return path has a real correctness bug (returning a temp dir that the deferred cleanup deletes) that I'd want fixed before merge, and the nested `if` is worth flattening while in there. With those addressed — ideally plus one wiring-level test — this is a good change that unblocks #1757 with a minimal diff.

**Comment by yxxhero** [comment]
Follow-up on the upstream-tracking point: opened helmfile/chartify#206 — `chartify crashes on empty render: expose sentinel error or treat as no-op`. Once that lands and helmfile bumps its chartify version, the `chartifyEmptyRenderOutputErrSubstring` match here can be retired. Linking it in the constant's doc comment (per the earlier suggestion) would keep the cleanup discoverable.

**Comment by ankit090701** [comment]
Thanks for the thorough review, @yxxhero — you're right, that was a real bug. Pushed a fix:

- **Blocking issue**: captured `originalChartPath` before the deps-rewrite reassigns `chartPath`, and return that from the no-op path instead. Verified this matters with a new end-to-end test (see below) that fails without the fix (asserting `os.Stat` on the returned path) and passes with it.
- Flattened the nested `if err != nil { if isChartifyEmptyRenderOutputError(err) { ... } }` into two sequential checks, per your suggestion.
- Linked https://github.com/helmfile/chartify/issues/206 in `chartifyEmptyRenderOutputErrSubstring`'s doc comment, so the text-match is easy to find and retire once that lands.
- Added `TestProcessChartification_EmptyRenderReturnsSurvivingPath`, which exercises the real `processChartification` → `chartify.Chartify` wiring (a local chart with a relative `file://` dependency, rendering zero resources, with a transformer configured) rather than just the pure `isChartifyEmptyRenderOutputError` helper. It asserts the returned path still exists on disk and that `Chart.yaml` is reachable from it — i.e. it directly catches the bug you found. I verified it fails on the old code and passes on the new code, and ran it against real `helm`+`kustomize` in a Linux container to confirm it's sound (matching how CI will run it).
  - One caveat: it's skipped on Windows (my local dev machine) — the `file://` dependency needed to trigger `rewriteChartDependencies`'s temp-copy path hits a separate, pre-existing Windows issue in chartify's dependency resolution (a drive letter embedded in a `file://` URL gets mis-joined with another path). That's unrelated to this fix; the test passes cleanly on Linux, which is what CI runs.

Merged in a trivial upstream `main` commit (an unrelated Actions version bump) to keep the branch current. All commits signed off per `CONTRIBUTING.md`.

File: pkg/state/state.go [hunk 1]
```diff
@@ -1954,6 +1974,27 @@ func (st *HelmState) processChartification(chartification *Chartify, release *Re
...
+// `--output-dir`, and an empty render leaves that directory empty, so the "must be the abs
+// path to the output directory" assertion fails on the resulting empty string. Chartify does
+// not expose a typed/sentinel error for this case, so we match on the error text; if this
+// substring ever stops matching a real chartify error, chartify's wording has changed and
+// this check needs to be revisited.
+//
+// Tracked upstream at https://github.com/helmfile/chartify/issues/206 - once chartify
+// exposes a sentinel error (or treats an empty render as a no-op itself) and this repo
+// bumps to that version, this text match can be retired.
+const chartifyEmptyRenderOutputErrSubstring = "it must be the abs path to the output directory"
```
[the remark this answers is not in the document]
Comment (ankit090701) [reply pkg/state/state.go]: Good catch, applied — verified against the actual chartify v0.28.0 source that `chartOutputDir` is formatted with `%q`, so the empty-render case produces exactly `unexpected dir entry "" it must be...`. Narrowed the constant to include that prefix, and added a test case asserting the same assertion text with a non-empty dir entry is correctly *not* matched. Re-ran the full suite including the real helm+kustomize integration test in a Linux container to confirm the narrower match still catches the actual bug.

**Review by yxxhero — APPROVED** [review]
